File: server/src/parser.ts
/**
 * WebRelease Template Language Parser
 */

export interface Position {
  line: number;
  character: number;
}

export interface Range {
  start: Position;
  end: Position;
}

export interface Diagnostic {
  range: Range;
  severity: number; // 1=Error, 2=Warning, 3=Information, 4=Hint
  message: string;
  code?: string;
}

// Functions that can only be called on single elements (not lists)
const SINGLE_ELEMENT_FUNCTIONS = new Set([
  'selectedValue',
  'selectedName',
  'selected'
]);

// Built-in functions
const BUILT_IN_FUNCTIONS = new Set([
  'pageTitle', 'currentTime', 'formatDate', 'isNull', 'isNotNull', 'isNumber',
  'number', 'string', 'divide', 'setScale', 'length', 'substring', 'indexOf',
  'contains', 'startsWith', 'endsWith', 'toUpperCase', 'toLowerCase', 'trim',
  'replace', 'split', 'join', 'round', 'floor', 'ceil', 'abs', 'min', 'max',
  'unsplit', 'generatePrice', 'generateBenefit', 'generateBakuage',
  'selectedValue', 'selectedName', 'selected'
]);

// Valid WebRelease tags
const VALID_TAGS = new Set([
  'wr-if', 'wr-then', 'wr-else', 'wr-switch', 'wr-case', 'wr-default',
  'wr-for', 'wr-break', 'wr-variable', 'wr-append', 'wr-clear',
  'wr-return', 'wr-error', 'wr-conditional', 'wr-cond', 'wr-comment'
]);

// Tag attributes
const TAG_ATTRIBUTES: Record<string, string[]> = {
  'wr-if': ['condition'],
  'wr-then': [],
  'wr-else': [],
  'wr-switch': ['value'],
  'wr-case': ['value'],
  'wr-default': [],
  'wr-for': ['list', 'string', 'times', 'variable', 'count', 'index'],
  'wr-break': ['condition'],
  'wr-variable': ['name', 'value'],
  'wr-append': ['name', 'value'],
  'wr-clear': ['name'],
  'wr-return': ['value'],
  'wr-error': ['condition', 'message'],
  'wr-conditional': ['condition'],
  'wr-cond': ['condition'],
  'wr-comment': [],
};

// Tags that don't require closing tags
const NO_CLOSE_TAGS = new Set([
  'wr-then', 'wr-else', 'wr-case', 'wr-default',
  'wr-variable', 'wr-append', 'wr-clear', 'wr-break',
  'wr-return', 'wr-error', 'wr-cond'
]);

/**
 * Expression Parser for WebRelease template expressions
 */
class ExpressionParser {
  private expression: string;
  private pos: number = 0;
  private listElements: Set<string>;
  private loopVariables: Set<string>;
  private errors: Array<{ position: number; message: string }> = [];

  constructor(
    expression: string,
    listElements: Set<string> = new Set(),
    loopVariables: Set<string> = new Set()
  ) {
    this.expression = expression.trim();
    this.listElements = listElements;
    this.loopVariables = loopVariables;
  }

  parse(): { isValid: boolean; errorMessage: string | null } {
    if (!this.expression) {
      return { isValid: true, errorMessage: null };
    }

    try {
      this.parseExpression();
      if (this.errors.length > 0) {
        return { isValid: false, errorMessage: this.errors[0].message };
      }
      return { isValid: true, errorMessage: null };
    } catch (e) {
      return { isValid: false, errorMessage: (e as Error).message };
    }
  }

  getErrors(): Array<{ position: number; message: string }> {
    return this.errors;
  }

  private parseExpression(): void {
    this.parseOrExpression();
  }

  private parseOrExpression(): void {
    this.parseAndExpression();
    while (this.matchOperator('||')) {
      this.parseAndExpression();
    }
  }

  private parseAndExpression(): void {
    this.parseComparisonExpression();
    while (this.matchOperator('&&')) {
      this.parseComparisonExpression();
    }
  }

  private parseComparisonExpression(): void {
    this.parseAdditiveExpression();
    while (this.matchAnyOperator(['==', '!=', '<=', '>=', '<', '>'])) {
      this.parseAdditiveExpression();
    }
  }

  private parseAdditiveExpression(): void {
    this.parseMultiplicativeExpression();
    while (this.matchAnyOperator(['+', '-'])) {
      this.parseMultiplicativeExpression();
    }
  }

  private parseMultiplicativeExpression(): void {
    this.parseUnaryExpression();
    while (this.matchAnyOperator(['*', '/', '%'])) {
      this.parseUnaryExpression();
    }
  }

  private parseUnaryExpression(): void {
    if (this.matchOperator('!')) {
      this.parseUnaryExpression();
    } else {
      this.parsePostfixExpression();
    }
  }

  private parsePostfixExpression(): void {
    const baseElement = this.parsePrimaryExpression();
    const elementChain: string[] = baseElement ? [baseElement] : [];

    while (true) {
      if (this.match('(')) {
        // Function call
        const funcName = elementChain.length > 0 ? elementChain[elementChain.length - 1] : null;
        this.parseFunctionArguments();
        this.expect(')');

        // Check if this is a single-element function called on a list
        if (funcName && SINGLE_ELEMENT_FUNCTIONS.has(funcName)) {
          if (elementChain.length > 1) {
            const elementPath = elementChain.slice(0, -1).join('.');
            this.checkListFunctionCall(elementPath, funcName);
          }
        }

        elementChain.length = 0; // Reset after function call
      } else if (this.match('[')) {
        // Array access
        this.parseExpression();
        this.expect(']');
        elementChain.length = 0; // Array access produces a single element
      } else if (this.match('.')) {
        // Member access
        const member = this.parseIdentifier();
        if (member) {
          elementChain.push(member);
        }
      } else {
        break;
      }
    }
  }

  private checkListFunctionCall(elementPath: string, funcName: string): void {
    // Check if the element path itself is a list element
    if (this.listElements.has(elementPath)) {
      this.errors.push({
        position: this.pos,
        message: `Cannot call '${funcName}()' on list element '${elementPath}'. Use a loop variable instead.`
      });
      return;
    }

    // Check if any prefix of the path is a list element
    const parts = elementPath.split('.');
    for (let i = 0; i < parts.length; i++) {
      const prefix = parts.slice(0, i + 1).join('.');
      if (this.listElements.has(prefix)) {
        this.errors.push({
          position: this.pos,
          message: `Cannot call '${funcName}()' on list element '${prefix}'. Use a loop variable to iterate over the list first.`
        });
        return;
      }
    }
  }

  private parsePrimaryExpression(): string | null {
    this.skipWhitespace();

    if (this.pos >= this.expression.length) {
      throw new Error('Unexpected end of expression');
    }

    // String literal
    if (this.expression[this.pos] === '"') {
      this.parseStringLiteral();
      return null;
    }
    // Number literal
    if (/\d/.test(this.expression[this.pos])) {
      this.parseNumberLiteral();
      return null;
    }
    // Parenthesized expression
    if (this.expression[this.pos] === '(') {
      this.match('(');
      this.parseExpression();
      this.expect(')');
      return null;
    }
    // Identifier
    return this.parseIdentifier();
  }

  private parseIdentifier(): string | null {
    this.skipWhitespace();

    if (this.pos >= this.expression.length) {
      return null;
    }

    if (!/[a-zA-Z_]/.test(this.expression[this.pos])) {
      return null;
    }

    const start = this.pos;
    while (this.pos < this.expression.length && /[a-zA-Z0-9_]/.test(this.expression[this.pos])) {
      this.pos++;
    }

    return this.expression.slice(start, this.pos);
  }

  private parseFunctionArguments(): void {
    this.skipWhitespace();

    if (this.pos < this.expression.length && this.expression[this.pos] !== ')') {
      this.parseExpression();

      while (this.match(',')) {
        this.parseExpression();
      }
    }
  }

  private parseStringLiteral(): void {
    this.expect('"');

    while (this.pos < this.expression.length && this.expression[this.pos] !== '"') {
      if (this.expression[this.pos] === '\\' && this.pos + 1 < this.expression.length) {
        this.pos += 2; // Skip escaped character
      } else {
        this.pos++;
      }
    }

    this.expect('"');
  }

  private parseNumberLiteral(): void {
    while (this.pos < this.expression.length && /[\d.]/.test(this.expression[this.pos])) {
      this.pos++;
    }
  }

  private match(ch: string): boolean {
    this.skipWhitespace();
    if (this.pos < this.expression.length && this.expression[this.pos] === ch) {
      this.pos++;
      return true;
    }
    return false;
  }

  private matchOperator(op: string): boolean {
    this.skipWhitespace();
    if (this.expression.slice(this.pos, this.pos + op.length) === op) {
      this.pos += op.length;
      return true;
    }
    return false;
  }

  private matchAnyOperator(ops: string[]): boolean {
    // Sort by length descending to match longest first
    const sortedOps = [...ops].sort((a, b) => b.length - a.length);
    for (const op of sortedOps) {
      if (this.matchOperator(op)) {
        return true;
      }
    }
    return false;
  }

  private expect(ch: string): void {
    this.skipWhitespace();
    if (this.pos >= this.expression.length || this.expression[this.pos] !== ch) {
      throw new Error(`Expected '${ch}' at position ${this.pos}`);
    }
    this.pos++;
  }

  private skipWhitespace(): void {
    while (this.pos < this.expression.length && /\s/.test(this.expression[this.pos])) {
      this.pos++;
    }
  }
}

/**
 * Main Template Parser
 */
export class TemplateParser {
  private content: string;
  private diagnostics: Diagnostic[] = [];
  private listElements: Set<string> = new Set();
  private loopVariables: Map<string, string> = new Map();

  constructor(content: string) {
    this.content = content;
  }

  parse(): Diagnostic[] {
    this.diagnostics = [];
    this.listElements = new Set();
    this.loopVariables = new Map();

    // First pass: collect list elements from wr-for tags
    this.collectListElements();

    // Validate expressions
    this.validateExpressions();

    // Validate tags
    this.validateTags();

    // Validate tag closures
    this.validateTagClosures();

    return this.diagnostics;
  }

  getListElements(): Set<string> {
    return this.listElements;
  }

  getLoopVariables(): Map<string, string> {
    return this.loopVariables;
  }

  private collectListElements(): void {
    const pattern = /<wr-for\s+([^>]*)>/g;
    let match;

    while ((match = pattern.exec(this.content)) !== null) {
      const attributesStr = match[1];

      // Extract list attribute
      const listMatch = /list=['"]([^'"]+)['"]/.exec(attributesStr);
      if (listMatch) {
        const listElement = listMatch[1];
        this.listElements.add(listElement);

        // Extract variable attribute
        const varMatch = /variable=['"]([^'"]+)['"]/.exec(attributesStr);
        if (varMatch) {
          this.loopVariables.set(varMatch[1], listElement);
        }
      }
    }
  }

  private validateExpressions(): void {
    const pattern = /%([^%]*)%/g;
    let match;

    while ((match = pattern.exec(this.content)) !== null) {
      const expression = match[1];
      const startPos = match.index;

      // Skip %% (escaped percent)
      if (expression === '') {
        continue;
      }

      // Parse expression with list element context
      const parser = new ExpressionParser(
        expression,
        this.listElements,
        new Set(this.loopVariables.keys())
      );
      const result = parser.parse();

      if (!result.isValid) {
        const { line, character } = this.getLineColumn(startPos);
        this.diagnostics.push({
          range: {
            start: { line, character },
            end: { line, character: character + expression.length + 2 }
          },
          severity: 1,
          message: `Expression error: ${result.errorMessage}`,
          code: 'expr-error'
        });
      }

      // Check for additional errors from the parser
      for (const error of parser.getErrors()) {
        const { line, character } = this.getLineColumn(startPos);
        this.diagnostics.push({
          range: {
            start: { line, character },
            end: { line, character: character + expression.length + 2 }
          },
          severity: 1,
          message: error.message,
          code: 'list-function-error'
        });
      }
    }
  }

  private validateTags(): void {
    const pattern = /<wr-(\w+)([^>]*)>/g;
    let match;

    while ((match = pattern.exec(this.content)) !== null) {
      const tagName = 'wr-' + match[1];
      const attributesStr = match[2];
      const startPos = match.index;

      if (!VALID_TAGS.has(tagName)) {
        const { line, character } = this.getLineColumn(startPos);
        this.diagnostics.push({
          range: {
            start: { line, character },
            end: { line, character: character + tagName.length + 2 }
          },
          severity: 1,
          message: `Unknown tag: ${tagName}`,
          code: 'unknown-tag'
        });
        continue;
      }

      // Validate attributes
      this.validateTagAttributes(tagName, attributesStr, startPos);
    }
  }

  private validateTagAttributes(tagName: string, attributesStr: string, startPos: number): void {
    // Match both single and double quoted attributes
    const attrPattern = /(\w+)='([^']*)'|(\w+)="([^"]*)"/g;
    let match;

    while ((match = attrPattern.exec(attributesStr)) !== null) {
      const attrName = match[1] || match[3];
      const attrValue = match[2] || match[4];

      // Check if attribute is valid for this tag
      const validAttrs = TAG_ATTRIBUTES[tagName] || [];
      if (!validAttrs.includes(attrName)) {
        const { line, character } = this.getLineColumn(startPos + match.index);
        this.diagnostics.push({
          range: {
            start: { line, character },
            end: { line, character: character + attrName.length }
          },
          severity: 2,
          message: `Unknown attribute '${attrName}' for tag '${tagName}'`,
          code: 'unknown-attribute'
        });
      }

      // Validate condition attribute
      if (attrName === 'condition' && attrValue) {
        const parser = new ExpressionParser(
          attrValue,
          this.listElements,
          new Set(this.loopVariables.keys())
        );
        const result = parser.parse();

        if (!result.isValid) {
          const { line, character } = this.getLineColumn(startPos);
          this.diagnostics.push({
            range: {
              start: { line, character },
              end: { line, character: character + attrValue.length }
            },
            severity: 1,
            message: `Condition syntax error: ${result.errorMessage}`,
            code: 'condition-syntax-error'
          });
        }

        // Check for list function errors
        for (const error of parser.getErrors()) {
          const { line, character } = this.getLineColumn(startPos);
          this.diagnostics.push({
            range: {
              start: { line, character },
              end: { line, character: character + attrValue.length }
            },
            severity: 1,
            message: error.message,
            code: 'list-function-error'
          });
        }
      }
    }
  }

  private validateTagClosures(): void {
    const openPattern = /<wr-(\w+)(?:\s|>)/g;
    const closePattern = /<\/wr-(\w+)>/g;

    // Collect all tag events with positions
    const events: Array<{ type: 'open' | 'close'; tagName: string; pos: number }> = [];

    let match;
    while ((match = openPattern.exec(this.content)) !== null) {
      events.push({ type: 'open', tagName: 'wr-' + match[1], pos: match.index });
    }

    while ((match = closePattern.exec(this.content)) !== null) {
      events.push({ type: 'close', tagName: 'wr-' + match[1], pos: match.index });
    }

    // Sort events by position
    events.sort((a, b) => a.pos - b.pos);

    // Stack-based validation
    const tagStack: Array<{ tagName: string; pos: number }> = [];

    for (const event of events) {
      if (event.type === 'open') {
        if (!NO_CLOSE_TAGS.has(event.tagName)) {
          tagStack.push({ tagName: event.tagName, pos: event.pos });
        }
      } else {
        if (NO_CLOSE_TAGS.has(event.tagName)) {
          continue;
        }

        // Find matching opening tag in stack
        let found = false;
        for (let i = tagStack.length - 1; i >= 0; i--) {
          if (tagStack[i].tagName === event.tagName) {
            tagStack.splice(i, 1);
            found = true;
            break;
          }
        }

        if (!found) {
          const { line, character } = this.getLineColumn(event.pos);
          this.diagnostics.push({
            range: {
              start: { line, character },
              end: { line, character: character + event.tagName.length + 3 }
            },
            severity: 1,
            message: `Closing tag '${event.tagName}' without matching opening tag`,
            code: 'unmatched-closing-tag'
          });
        }
      }
    }

    // Check for unclosed tags
    for (const { tagName, pos } of tagStack) {
      const { line, character } = this.getLineColumn(pos);
      this.diagnostics.push({
        range: {
          start: { line, character },
          end: { line, character: character + tagName.length + 2 }
        },
        severity: 1,
        message: `Unclosed tag '${tagName}'`,
        code: 'unclosed-tag'
      });
    }
  }

  private getLineColumn(pos: number): { line: number; character: number } {
    let line = 0;
    let character = 0;

    for (let i = 0; i < Math.min(pos, this.content.length); i++) {
      if (this.content[i] === '\n') {
        line++;
        character = 0;
      } else {
        character++;
      }
    }

    return { line, character };
  }
}

// Export completion items
export const TAG_COMPLETIONS = Array.from(VALID_TAGS).map(tag => ({
  label: tag,
  kind: 14, // Keyword
  detail: `WebRelease tag: ${tag}`,
  documentation: getTagDocumentation(tag)
}));

export const FUNCTION_COMPLETIONS = Array.from(BUILT_IN_FUNCTIONS).map(func => ({
  label: func,
  kind: 3, // Function
  detail: `WebRelease function: ${func}()`,
  insertText: `${func}()`
}));

function getTagDocumentation(tag: string): string {
  const docs: Record<string, string> = {
    'wr-if': 'Conditional rendering. Use with wr-then and wr-else.',
    'wr-then': 'Content to render when wr-if condition is true.',
    'wr-else': 'Content to render when wr-if condition is false.',
    'wr-for': 'Loop over a list, string, or number of times.',
    'wr-switch': 'Switch statement for multiple conditions.',
    'wr-case': 'Case within a wr-switch block.',
    'wr-default': 'Default case within a wr-switch block.',
    'wr-variable': 'Define a variable.',
    'wr-append': 'Append value to a variable.',
    'wr-clear': 'Clear a variable.',
    'wr-break': 'Break out of a loop.',
    'wr-return': 'Return a value.',
    'wr-error': 'Raise an error.',
    'wr-conditional': 'Conditional block.',
    'wr-cond': 'Condition within wr-conditional.',
    'wr-comment': 'Comment block (not rendered).'
  };
  return docs[tag] || '';
}

File: server/parser.py
"""
WebRelease Template Language Parser - Version 2
Fixed tag closure validation
"""

import re
from typing import List, Dict, Tuple, Optional, Any
from dataclasses import dataclass
from enum import Enum


class TokenType(Enum):
    """Token types for WebRelease template language"""
    TEXT = "TEXT"
    PERCENT = "PERCENT"
    TAG_OPEN = "TAG_OPEN"
    TAG_CLOSE = "TAG_CLOSE"
    TAG_NAME = "TAG_NAME"
    ATTRIBUTE_NAME = "ATTRIBUTE_NAME"
    ATTRIBUTE_VALUE = "ATTRIBUTE_VALUE"
    IDENTIFIER = "IDENTIFIER"
    NUMBER = "NUMBER"
    STRING = "STRING"
    OPERATOR = "OPERATOR"
    LPAREN = "LPAREN"
    RPAREN = "RPAREN"
    LBRACKET = "LBRACKET"
    RBRACKET = "RBRACKET"
    DOT = "DOT"
    COMMA = "COMMA"
    EOF = "EOF"
    ERROR = "ERROR"


@dataclass
class Token:
    """Represents a token in the template"""
    type: TokenType
    value: str
    line: int
    column: int
    length: int


@dataclass
class Position:
    """Represents a position in the document"""
    line: int
    character: int


@dataclass
class Range:
    """Represents a range in the document"""
    start: Position
    end: Position


@dataclass
class Diagnostic:
    """Represents a diagnostic message"""
    range: Range
    severity: int  # 1=Error, 2=Warning, 3=Information, 4=Hint
    message: str
    code: Optional[str] = None


class ExpressionParser:
    """Parses WebRelease expressions"""
    
    # Built-in functions
    BUILT_IN_FUNCTIONS = {
        'pageTitle', 'currentTime', 'formatDate', 'isNull', 'isNotNull', 'isNumber',
        'number', 'string', 'divide', 'setScale', 'length', 'substring', 'indexOf',
        'contains', 'startsWith', 'endsWith', 'toUpperCase', 'toLowerCase', 'trim',
        'replace', 'split', 'join', 'round', 'floor', 'ceil', 'abs', 'min', 'max'
    }
    
    # Operators
    OPERATORS = {
        '==', '!=', '<', '<=', '>', '>=', '&&', '||', '!',
        '+', '-', '*', '/', '%'
    }
    
    def __init__(self, expression: str):
        self.expression = expression.strip()
        self.pos = 0
    
    def parse(self) -> Tuple[bool, Optional[str]]:
        """
        Parse expression and return (is_valid, error_message)
        """
        if not self.expression:
            return True, None
        
        try:
            self._parse_expression()
            return True, None
        except Exception as e:
            return False, str(e)
    
    def _parse_expression(self):
        """Parse a complete expression"""
        self._parse_or_expression()
    
    def _parse_or_expression(self):
        """Parse logical OR expression"""
        self._parse_and_expression()
        while self._match_operator('||'):
            self._parse_and_expression()
    
    def _parse_and_expression(self):
        """Parse logical AND expression"""
        self._parse_comparison_expression()
        while self._match_operator('&&'):
            self._parse_comparison_expression()
    
    def _parse_comparison_expression(self):
        """Parse comparison expression"""
        self._parse_additive_expression()
        while self._match_any_operator(['==', '!=', '<', '<=', '>', '>=']):
            self._parse_additive_expression()
    
    def _parse_additive_expression(self):
        """Parse addition/subtraction expression"""
        self._parse_multiplicative_expression()
        while self._match_any_operator(['+', '-']):
            self._parse_multiplicative_expression()
    
    def _parse_multiplicative_expression(self):
        """Parse multiplication/division expression"""
        self._parse_unary_expression()
        while self._match_any_operator(['*', '/', '%']):
            self._parse_unary_expression()
    
    def _parse_unary_expression(self):
        """Parse unary expression"""
        if self._match_operator('!'):
            self._parse_unary_expression()
        else:
            self._parse_postfix_expression()
    
    def _parse_postfix_expression(self):
        """Parse postfix expression (function calls, array access, member access)"""
        self._parse_primary_expression()
        
        while True:
            if self._match('('):
                # Function call
                self._parse_function_arguments()
                self._expect(')')
            elif self._match('['):
                # Array access
                self._parse_expression()
                self._expect(']')
            elif self._match('.'):
                # Member access
                self._parse_identifier()
            else:
                break
    
    def _parse_primary_expression(self):
        """Parse primary expression"""
        self._skip_whitespace()
        
        if self.pos >= len(self.expression):
            raise ValueError("Unexpected end of expression")
        
        # String literal
        if self.expression[self.pos] == '"':
            self._parse_string_literal()
        # Number literal
        elif self.expression[self.pos].isdigit():
            self._parse_number_literal()
        # Parenthesized expression
        elif self.expression[self.pos] == '(':
            self._match('(')
            self._parse_expression()
            self._expect(')')
        # Identifier (element reference or function call)
        else:
            self._parse_identifier()
    
    def _parse_identifier(self):
        """Parse identifier"""
        self._skip_whitespace()
        
        if self.pos >= len(self.expression) or not (self.expression[self.pos].isalnum() or self.expression[self.pos] in '_'):
            raise ValueError(f"Expected identifier at position {self.pos}")
        
        start = self.pos
        while self.pos < len(self.expression) and (self.expression[self.pos].isalnum() or self.expression[self.pos] in '_'):
            self.pos += 1
        
        identifier = self.expression[start:self.pos]
        return identifier
    
    def _parse_function_arguments(self):
        """Parse function arguments"""
        self._skip_whitespace()
        
        if self.pos < len(self.expression) and self.expression[self.pos] != ')':
            self._parse_expression()
            
            while self._match(','):
                self._parse_expression()
    
    def _parse_string_literal(self):
        """Parse string literal"""
        self._expect('"')
        
        while self.pos < len(self.expression) and self.expression[self.pos] != '"':
            if self.expression[self.pos] == '\\' and self.pos + 1 < len(self.expression):
                self.pos += 2  # Skip escaped character
            else:
                self.pos += 1
        
        self._expect('"')
    
    def _parse_number_literal(self):
        """Parse number literal"""
        while self.pos < len(self.expression) and (self.expression[self.pos].isdigit() or self.expression[self.pos] == '.'):
            self.pos += 1
    
    def _match(self, ch: str) -> bool:
        """Match a single character"""
        self._skip_whitespace()
        if self.pos < len(self.expression) and self.expression[self.pos] == ch:
            self.pos += 1
            return True
        return False
    
    def _match_operator(self, op: str) -> bool:
        """Match an operator"""
        self._skip_whitespace()
        if self.expression[self.pos:self.pos+len(op)] == op:
            self.pos += len(op)
            return True
        return False
    
    def _match_any_operator(self, ops: List[str]) -> bool:
        """Match any of the given operators"""
        for op in sorted(ops, key=len, reverse=True):  # Match longest first
            if self._match_operator(op):
                return True
        return False
    
    def _expect(self, ch: str):
        """Expect a character"""
        self._skip_whitespace()
        if self.pos >= len(self.expression) or self.expression[self.pos] != ch:
            raise ValueError(f"Expected '{ch}' at position {self.pos}")
        self.pos += 1
    
    def _skip_whitespace(self):
        """Skip whitespace"""
        while self.pos < len(self.expression) and self.expression[self.pos].isspace():
            self.pos += 1


class TemplateParser:
    """Main parser for WebRelease templates"""
    
    # Valid WebRelease tags
    VALID_TAGS = {
        'wr-if', 'wr-then', 'wr-else', 'wr-switch', 'wr-case', 'wr-default',
        'wr-for', 'wr-break', 'wr-variable', 'wr-append', 'wr-clear',
        'wr-return', 'wr-error', 'wr-conditional', 'wr-cond', 'wr-comment'
    }
    
    # Tag attributes
    TAG_ATTRIBUTES = {
        'wr-if': ['condition'],
        'wr-then': [],
        'wr-else': [],
        'wr-switch': ['value'],
        'wr-case': ['value'],
        'wr-default': [],
        'wr-for': ['list', 'string', 'times', 'variable', 'count', 'index'],
        'wr-break': ['condition'],
        'wr-variable': ['name', 'value'],
        'wr-append': ['name', 'value'],
        'wr-clear': ['name'],
        'wr-return': ['value'],
        'wr-error': ['condition', 'message'],
        'wr-conditional': ['condition'],
        'wr-cond': ['condition'],
        'wr-comment': [],
    }
    
    # Tags that don't require closing tags (or have optional closing)
    NO_CLOSE_TAGS = {'wr-then', 'wr-else', 'wr-case', 'wr-default'}
    
    def __init__(self, content: str):
        self.content = content
        self.diagnostics: List[Diagnostic] = []
        self.elements: Dict[str, str] = {}  # Element name -> type mapping
    
    def parse(self) -> List[Diagnostic]:
        """Parse the template and return diagnostics"""
        self.diagnostics = []
        
        # Find all expressions
        self._validate_expressions()
        
        # Find all tags
        self._validate_tags()
        
        # Find all tag closures
        self._validate_tag_closures()
        
        return self.diagnostics
    
    def _validate_expressions(self):
        """Validate all expressions in the template"""
        # Find all %...% expressions
        pattern = r'%([^%]*)%'
        
        for match in re.finditer(pattern, self.content):
            expression = match.group(1)
            start_pos = match.start()
            
            # Skip %% (escaped percent)
            if expression == '':
                continue
            
            # Parse expression
            parser = ExpressionParser(expression)
            is_valid, error_msg = parser.parse()
            
            if not is_valid:
                line, column = self._get_line_column(start_pos)
                diagnostic = Diagnostic(
                    range=Range(
                        start=Position(line, column),
                        end=Position(line, column + len(expression) + 2)
                    ),
                    severity=1,  # Error
                    message=f"Expression syntax error: {error_msg}",
                    code="expr-syntax-error"
                )
                self.diagnostics.append(diagnostic)
    
    def _validate_tags(self):
        """Validate all WebRelease tags"""
        # Find all <wr-...> tags
        pattern = r'<wr-(\w+)([^>]*)>'
        
        for match in re.finditer(pattern, self.content):
            tag_name = 'wr-' + match.group(1)
            attributes_str = match.group(2)
            start_pos = match.start()
            
            if tag_name not in self.VALID_TAGS:
                line, column = self._get_line_column(start_pos)
                diagnostic = Diagnostic(
                    range=Range(
                        start=Position(line, column),
                        end=Position(line, column + len(tag_name) + 2)
                    ),
                    severity=1,  # Error
                    message=f"Unknown tag: {tag_name}",
                    code="unknown-tag"
                )
                self.diagnostics.append(diagnostic)
                continue
            
            # Validate attributes
            self._validate_tag_attributes(tag_name, attributes_str, start_pos)
    
    def _validate_tag_attributes(self, tag_name: str, attributes_str: str, start_pos: int):
        """Validate tag attributes"""
        # Parse attributes
        attr_pattern = r'(\w+)="([^"]*)"'
        
        found_attrs = set()
        for match in re.finditer(attr_pattern, attributes_str):
            attr_name = match.group(1)
            attr_value = match.group(2)
            found_attrs.add(attr_name)
            
            # Check if attribute is valid for this tag
            valid_attrs = self.TAG_ATTRIBUTES.get(tag_name, [])
            if attr_name not in valid_attrs:
                line, column = self._get_line_column(start_pos + match.start())
                diagnostic = Diagnostic(
                    range=Range(
                        start=Position(line, column),
                        end=Position(line, column + len(attr_name))
                    ),
                    severity=2,  # Warning
                    message=f"Unknown attribute '{attr_name}' for tag '{tag_name}'",
                    code="unknown-attribute"
                )
                self.diagnostics.append(diagnostic)
            
            # Validate attribute value (if it's a condition or expression)
            if attr_name == 'condition' and attr_value:
                parser = ExpressionParser(attr_value)
                is_valid, error_msg = parser.parse()
                if not is_valid:
                    line, column = self._get_line_column(start_pos + match.start(2))
                    diagnostic = Diagnostic(
                        range=Range(
                            start=Position(line, column),
                            end=Position(line, column + len(attr_value))
                        ),
                        severity=1,  # Error
                        message=f"Condition syntax error: {error_msg}",
                        code="condition-syntax-error"
                    )
                    self.diagnostics.append(diagnostic)
    
    def _validate_tag_closures(self):
        """Validate that tags are properly closed"""
        # Find all opening and closing tags
        open_pattern = r'<wr-(\w+)(?:\s|>)'
        close_pattern = r'</wr-(\w+)>'
        
        # Simple stack-based validation
        tag_stack = []
        
        for match in re.finditer(open_pattern, self.content):
            tag_name = 'wr-' + match.group(1)
            # Don't push tags that don't need closing
            if tag_name not in self.NO_CLOSE_TAGS:
                tag_stack.append((tag_name, match.start()))
        
        for match in re.finditer(close_pattern, self.content):
            tag_name = 'wr-' + match.group(1)
            # Skip closing tags that don't need opening
            if tag_name in self.NO_CLOSE_TAGS:
                continue
            
            if tag_stack and tag_stack[-1][0] == tag_name:
                tag_stack.pop()
            else:
                line, column = self._get_line_column(match.start())
                diagnostic = Diagnostic(
                    range=Range(
                        start=Position(line, column),
                        end=Position(line, column + len(tag_name) + 3)
                    ),
                    severity=1,  # Error
                    message=f"Closing tag '{tag_name}' without matching opening tag",
                    code="unmatched-closing-tag"
                )
                self.diagnostics.append(diagnostic)
        
        # Check for unclosed tags
        for tag_name, start_pos in tag_stack:
            line, column = self._get_line_column(start_pos)
            diagnostic = Diagnostic(
                range=Range(
                    start=Position(line, column),
                    end=Position(line, column + len(tag_name) + 2)
                ),
                severity=1,  # Error
                message=f"Unclosed tag '{tag_name}'",
                code="unclosed-tag"
            )
            self.diagnostics.append(diagnostic)
    
    def _get_line_column(self, pos: int) -> Tuple[int, int]:
        """Get line and column from position"""
        line = 0
        column = 0
        
        for i in range(min(pos, len(self.content))):
            if self.content[i] == '\n':
                line += 1
                column = 0
            else:
                column += 1
        
        return line, column
